File: lags/ruby/lags.rb
def optimal_price(flights)
  depart = []
  duree  = []
  prices = []
  arrive = 0
  best_price = 0
  gainmax = []
  i = 0
  j = 0

  flights.each do |flight|
    depart << flight[0]
    duree  << flight[1]
    prices << flight[flight.length - 1]
  end

  if flights.length > 1
      while i < flights.length-1 do
          arrive = flights[i][0] + flights[i][1]
          while j < flights.length-1 do
              if arrive <= depart[j+1]
                  gainmax << prices[i] + prices[j+1]
                  best_price = gainmax.max
              else
                  best_price = prices.max
              end
                j +=1
          end
        i += 1
      end
  else
      best_price = prices.max
  end

  best_price
end

File: lags/python/lags.py
def lags(demandes):
    gainmax = 0
    j = 0

    for i in range(len(demandes)):
        demande = demandes[i]

        while j < len(demandes) - 1:
            demande_suivante = demandes[j + 1]
            if demande.compatible(demande_suivante):
                gainmax = demande.prix + demande_suivante.prix
            j = j + 1

        # if i < len(demandes) - 1:
        #     demande_suivante = demandes[i + 1]
        #     if demande.compatible(demande_suivante):
        #         gainmax = demande.prix + demande_suivante.prix

        if demande.prix > gainmax:
            gainmax = demande.prix

    return gainmax

class Demande:
    def __init__(self, depart, duree, prix):
        self.depart = depart
        self.duree = duree
        self.prix = prix
        self.arrive = depart + duree

    def compatible(self, autre_demande):
        return self.arrive <= autre_demande.depart